app/src/main/java/org/thoughtcrime/securesms/conversation/v2/AttachmentDownloadHelper.kt

```diff
@@ -0,0 +1,120 @@
+import kotlinx.coroutines.channels.Channel.Factory.UNLIMITED
+import kotlinx.coroutines.flow.flowOn
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.flow.receiveAsFlow
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.plus
+import kotlinx.coroutines.withContext
+import org.session.libsession.database.StorageProtocol
+import org.session.libsession.messaging.jobs.AttachmentDownloadJob
+import org.session.libsession.messaging.jobs.AttachmentUploadJob
+import org.session.libsession.messaging.jobs.JobQueue
+import org.session.libsession.messaging.sending_receiving.attachments.AttachmentTransferProgress
+import org.session.libsession.messaging.sending_receiving.attachments.DatabaseAttachment
+import org.session.libsignal.utilities.Log
+import org.thoughtcrime.securesms.database.MmsDatabase
+import org.thoughtcrime.securesms.database.model.MessageRecord
+import org.thoughtcrime.securesms.util.flatten
+import org.thoughtcrime.securesms.util.timedBuffer
+
+class AttachmentDownloadHelper(
+    private val storage: StorageProtocol,
+    mmsDatabase: MmsDatabase,
+    jobQueue: JobQueue = JobQueue.shared,
+    scope: CoroutineScope = CoroutineScope(Dispatchers.Default) + SupervisorJob(),
+) {
+    companion object {
+        private const val BUFFER_TIMEOUT_MILLS = 500L
+        private const val BUFFER_MAX_ITEMS = 10
+        private const val LOG_TAG = "AttachmentDownloadHelper"
+    }
+
+    private val downloadRequests = Channel<DatabaseAttachment>(UNLIMITED)
+
+    init {
+        scope.launch(Dispatchers.Default) {
+            downloadRequests
+                .receiveAsFlow()
+                .timedBuffer(BUFFER_TIMEOUT_MILLS, BUFFER_MAX_ITEMS)
+                .map { attachments ->
+                    val pendingAttachmentIDs = storage
+                        .getAllPendingJobs(AttachmentDownloadJob.KEY, AttachmentUploadJob.KEY)
+                        .values
+                        .mapNotNullTo(hashSetOf()) {
+                            (it as? AttachmentUploadJob)?.attachmentID
+                                ?: (it as? AttachmentDownloadJob)?.attachmentID
+                        }
+
+                    val messagesByID = mmsDatabase.getMessages(attachments.map { it.mmsId })
+                        .associateBy { it.id }
+
+                    // Before handling out attachment to the download task, we need to
+                    // check the requisite for that attachment. This check is very likely to be
+                    // performed again in the download task, but adding stuff into job system
+                    // is expensive so we need to avoid spawning new task whenever we can.
+                    attachments.filter { attachment ->
+                        attachment.attachmentId.rowId !in pendingAttachmentIDs &&
+                                eligibleForDownloadTask(attachment, messagesByID[attachment.mmsId])
+                    }
+                }
+                .flowOn(Dispatchers.IO)
```

Comment:
so we `flowOn(IO)` here which applies to upstream, but launch on default up there, applying to downstream.

I guess flowOn wins on this stream, but why 1. specify default up there, and 2. why wouldn't it be the same IO if it must be specified. or 3. can we just specify up there alone?

Reply:
It's usually good to start in default ( we could even remove it up there as it's redundant ), and leave the responsibility of contexts for the blocks that need it ( usually a suspend function somewhere ). In this case it is applying it to the code in the mapping, then the collection further down is back on the default thread. It wouldn't if you forced IO at the top.

Reply:
Let me put the dispatchers to where they really belong. The `flowOn` is very ambiguous and should be discouraged

Reply:
I believe there is no need to specify dispatchers anywhere in here, as the `Dispatchers.DEFAULT` dictates where all the flow things will execute on. As long as none of the stuff in this class is run on main thread it's all good.